packages/flutter_tools/lib/src/runner/flutter_command_runner.dart

```diff
@@ -263,10 +223,10 @@ class FlutterCommandRunner extends CommandRunner<void> {
       FlutterTesterDevices.showFlutterTesterDevice = true;
     }
 
-    // We must set Cache.flutterRoot early because other features use it (e.g.
-    // enginePath's initializer uses it).
-    final String flutterRoot = topLevelResults['flutter-root'] as String ?? defaultFlutterRoot;
-    Cache.flutterRoot = globals.fs.path.normalize(globals.fs.path.absolute(flutterRoot));
+    if (topLevelResults.wasParsed('flutter-root')) {
+      Cache.flutterRoot = globals.fs.path.normalize(
+        globals.fs.path.absolute(topLevelResults['flutter-root'] as String));
```

Comment:
I think the flutter-root argument should be deprecated, because it massively complicates startup. The cache flutter root needs to be initialized before many of our interfaces are created/accessed for the first time, but arg parsing can only be done after a command has executed.

Since this is a fairly advanced use case, replacing it with the environment variable seems like a reasonable compromise.